fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/filters/HttpMethodFilter.java

```diff
@@ -0,0 +1,81 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.infrastructure.core.filters;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.ext.Provider;
+import org.apache.fineract.infrastructure.core.config.FineractProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.GenericFilterBean;
+
+@Provider
+@Component
+@Scope("singleton")
+public class HttpMethodFilter extends GenericFilterBean {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HttpMethodFilter.class);
+
+    private final String DEFAULT_HTTP_METHODS_READ = "GET";
+    private final String DEFAULT_HTTP_METHODS_WRITE = "POST,PUT,DELETE,OPTIONS";
+    private final String DEFAULT_HTTP_METHODS_ALL = "GET,POST,PUT,DELETE,OPTIONS";
+
+    private final String fineractInstanceType;
+    private final List<String> httpMethodsAllowedList;
+
+    @Autowired
+    public HttpMethodFilter(final FineractProperties fineractProperties) {
+        if (fineractProperties.getMode().isReadOnlyMode()) {
+            this.fineractInstanceType = "Read";
+            this.httpMethodsAllowedList = Arrays.asList(DEFAULT_HTTP_METHODS_READ.toUpperCase().split(","));
+        } else if (fineractProperties.getMode().isWriteEnabled()) {
+            this.fineractInstanceType = "Write";
+            this.httpMethodsAllowedList = Arrays.asList(DEFAULT_HTTP_METHODS_WRITE.toUpperCase().split(","));
+        } else {
+            this.fineractInstanceType = "Read/Write";
+            this.httpMethodsAllowedList = Arrays.asList(DEFAULT_HTTP_METHODS_ALL.toUpperCase().split(","));
+        }
+        LOG.info("Fineract Instance type: {} {} methods", this.fineractInstanceType, this.httpMethodsAllowedList.size());
+    }
+
+    @Override
+    public void doFilter(final ServletRequest req, final ServletResponse res, final FilterChain chain)
```

Comment:
A pseudo code for this doFilter would look the following:
```
doFilter {
  if (isGetMethod() && isReadInstance()) {
    proceed()
  } else if (isAnythingElse() && isWriteInstance()) {
    proceed()
  } else {
    error()
  }
}
```
And later on if we need an exception list we could have a list of API paths like:
```
Set<String> EXCEPTION_API_PATHS = Sets.of(
    "/loans", ...) // just as an example of course
```
Then the doFilter's beginning gets extended with
```
if (EXCEPTION_API_PATHS.contains(currentApiPath)) {
   proceed()
}
```

Reply:
Done!